proof/refine/ARM/Schedule_R.thy

```diff
@@ -712,8 +712,13 @@ lemma tcbSchedDequeue_invs'[wp]:
   done
 
 lemma cur_thread_update_corres:
-  "corres dc \<top> \<top> (modify (cur_thread_update (\<lambda>_. t))) (setCurThread t)"
+  "corres dc (pspace_aligned and pspace_distinct and valid_ready_qs) \<top>
```

Comment:
I'd like to test my understanding. Can you confirm?

1. You want to make valid_ready_queues' simpler by removing the runnable condition.
2. Because you were using the runnable condition in some hoare logic proofs, you need to add an assertion to continue using this invariant in those places.
3. Because there is now an extra assertion in the haskell, you need to show it holds in the corresponding corres proofs.

Thus, while you make it easier to do hoare logic with valid_ready_queues', you make some corres proofs a little harder. Is that right?

Reply:
(I think this is a good way forward and really like the approach in general, just trying to understand why this corres proof now requires stronger preconditions.)

Reply:
Yes, exactly! 

The corres proofs are a little harder. Putting the assert at the beginning of functions makes it easy to use the assert in Hoare logic proofs by using `hoare_seq_ext[OF _ stateAssert_sp]` to grab the assertion. So the Hoare logic poofs generally don't change much (often they are easier in some spots because the postcondition is weaker).

The little method `add_ready_qs_runnable` together with the rule `corres_stateAssert_add_assumption` is really all you need. So 3 lines or so per assertion per corres proof (plus the short cross lemma and the new method)